Split Http2MultiplexCodec into Frame- and MultiplexCodec + Tests.

buchgr opened the pull request:
Some comments regarding my changes:

- The `Http2StreamFrames` only have the `stream id` attached, as opposed to a `Http2Stream` object. I think that makes more sense, as this cleanly separates the new h2childchan API from the callback-based API. Also, many methods of `Http2Stream` don't make sense and are not supposed to be called from within a child channel. If we think that some information is useful within a child channel, I propose to add a `Http2StreamChannel` interface with e.g. the stream state, etc. attached, so that a user can cast the channel e.g. `((Http2StreamChannel) ctx.channel()).streamState()`. Also, testing gets easier as the `Http2MultiplexCodec` tests don't need to mock a `Http2Stream` object.
- The `Http2MultiplexCodec` uses a hashmap to manage the `Http2StreamChannels`, as opposed to using a `PropertyKey`. I chose that also to cleanly separate the new API from the exisiting API.

Some questions regarding my changes:

- ~~Should we have the `Http2FrameCodec` extend from the `Http2ConnectionHandler`? The constructor of the `Http2ConnectionHandler` makes this kinda messy (we could provide a factory or so). The argument for doing so, would be to have fewer channel handlers in the pipeline. I am not sure if we should worry about that at this stage though. Thoughts?~~

---------------------------------------------------------------------------------------------------------
Fixes #4914

Motivation:

Quote from issue 4914:
"Http2MultiplexCodec currently does two things: mapping the existing h2 API to frames and managing the child channels.

It would be better if the two parts were separated. This would allow less-coupled development of the HTTP/2 handlers (flow control could be its own handler, for instance) and allow applications to insert themselves between all streams and the codec, which permits custom logic and could be used, in part, to implement custom frame types.

It would also greatly ease testing, as the child channel could be tested by itself without dealing with how frames are encoded on the wire."

Modifications:

- Split the Http2MultiplexCodec into Http2FrameCodec and Http2MultiplexCodec. The Http2FrameCodec interacts with the existing HTTP/2 callback-based API, while the Http2MulitplexCodec is completely independent of it and simply multiplexes Http2StreamFrames to the child channels. Additionally, the Http2Codec handler is introduced, which is a convenience class that simply sets up the Http2FrameCodec and Http2MultiplexCodec in the channel pipeline and removes itself.

- Improved test coverage quite a bit.

Result:

- The original Http2MultiplexCodec is split into Http2FrameCodec and Http2MultiplexCodec.
- More tests for higher confidence in the code.

buchgr commented:
@ejona86 @nmittler @Scottmitch

buchgr commented on codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2GoAwayFrame.java:
```diff
@@ -87,6 +93,17 @@ public Http2GoAwayFrame setExtraStreamIds(int extraStreamIds) {
+
+    @Override
+    public Integer lastStreamId() {
+        return lastStreamId;
```
@ejona86 
I would prefer to not have the `last stream id`, as I think `extraStreamIds` is more user friendly ... one should not care about the exact stream id. However, when receiving a `GOAWAY` frame, the `Http2FrameCodec` needs to communicate to the `Http2MultiplexCodec` which streams the `GOAWAY` applies to. Any good ideas on how else to do this?

ejona86 replied:
I think I was going to go with something more like a List<Http2Stream>. That way there wouldn't be any need for a second `int → stream` mapping in the multiplexing handler. It isn't without problems, but seemed simpler than keeping another map (as a map may require every handler between the multiplexing codec and the frame codec to maintain its own map).

ejona86 replied:
I was thinking later it would be fairly easy to make a new Http2DifferentStream in the frame handler and convert 1:1. We're mainly really only gaining private properties on Http2Stream, but that is used in both directions multiplex codec -> frame codec and frame codec -> multiplex codec.

Although, if we're talking about having a map on each stream for properties (likely one per handler) or a map of streams for properties (likely one per handler), I guess it doesn't matter much which we go with. Having a map on each stream is more predictable, so we can use the array-based indexing, and doesn't require noticing stream death. Important properties could also become first-class fields (like maybe the child channel). But both approaches do seem a bit wonky, so I'm open to suggestions and on the fence.

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/AbstractHttp2StreamStateEvent.java:
```diff
@@ -0,0 +1,30 @@
+ */
+package io.netty.handler.codec.http2;
+
+public class AbstractHttp2StreamStateEvent implements Http2StreamStateEvent {
```
If the class name is `Abstract*` we should also make it abstract + the constructor protected

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2GoAwayFrame.java:
```diff
@@ -87,6 +93,17 @@ public Http2GoAwayFrame setExtraStreamIds(int extraStreamIds) {
+    }
+
+    @Override
+    public Integer lastStreamId() {
```
It looks a bit "inconsistent" to have the set method take an int and the get method return an Integer.

buchgr replied:
@normanmaurer yeah the intention was to return `null` if not yet set. Maybe more consistent with other code, if we returned `-1` instead. Hopefully in this code review we will come up with a away to completely remove `lastStreamId` (see https://github.com/netty/netty/pull/5324#discussion_r65067122).

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Codec.java:
```diff
@@ -0,0 +1,107 @@
+        this(server, streamHandler, streamGroup, new DefaultHttp2FrameWriter());
+    }
+
+    Http2Codec(boolean server, ChannelHandler streamHandler,
```
private ?

buchgr replied:
@normanmaurer once I ll add tests to `Http2CodecTest` 😅, this constructor will be needed there. Should I add a comment `// Visible for testing`?

normanmaurer replied:
Yes

> Am 30.05.2016 um 16:15 schrieb Jakob Buchgraber <notifications@github.com>:
> 
> In codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Codec.java:
> 
> > +    }
> > +
> > +    /**
> > +     * Construct a new handler whose child channels run in a different event loop.
> > +     *
> > +     * @param server {@code true} this is a server
> > +     * @param streamHandler the handler added to channels for remotely-created streams. It must be
> > +     *     {@link ChannelHandler.Sharable}.
> > +     * @param streamGroup event loop for registering child channels
> > +     */
> > +    public Http2Codec(boolean server, ChannelHandler streamHandler,
> > +                      EventLoopGroup streamGroup) {
> > +        this(server, streamHandler, streamGroup, new DefaultHttp2FrameWriter());
> > +    }
> > +
> > +    Http2Codec(boolean server, ChannelHandler streamHandler,
> @normanmaurer once I ll add tests to Http2CodecTest 😅, this constructor will be needed there. Should I add a comment // Visible for testing?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Codec.java:
```diff
@@ -0,0 +1,107 @@
+ * ChannelConfig#setAutoRead(boolean)} are supported.
+ */
+@UnstableApi
+public final class Http2Codec extends ChannelDuplexHandler {
```
Why not have this extends `CombinedChannelDuplexHandler` ?

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Frame.java:
```diff
@@ -19,4 +19,8 @@
-public interface Http2Frame { }
+public interface Http2Frame {
+
+    String name();
```
javadocs

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FrameCodec.java:
```diff
@@ -0,0 +1,266 @@
+            upgrade.upgradeRequest().headers().setInt(
+                    HttpConversionUtil.ExtensionHeaderNames.STREAM_ID.text(), Http2CodecUtil.HTTP_UPGRADE_STREAM_ID);
+            new InboundHttpToHttp2Adapter(http2Handler.connection(), http2Handler.decoder().frameListener())
+                    .channelRead(ctx, upgrade.upgradeRequest().retain());
```
I don't think we need to call retain again here as we already did above.

buchgr replied:
@normanmaurer I didn't write this code, but I think we do. The retain on line 88 is for the `UpgradeEvent` and the one on line 98 for the `UpgradeRequest`.

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2GoAwayFrame.java:
```diff
@@ -46,6 +46,13 @@
+     */
+    Integer lastStreamId();
+
+    Http2GoAwayFrame setLastStreamId(int lastStreamId);
```
javdocs

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java:
```diff
@@ -28,163 +23,155 @@
+            Http2StreamFrame frame = (Http2StreamFrame) msg;
+            int streamId = frame.streamId();
+            Http2StreamChannel childChannel = childChannels.get(streamId);
+            if (childChannel == null) {
```
do we need to release the fame in this case ?

buchgr replied:
@normanmaurer done.

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java:
```diff
@@ -28,163 +23,155 @@
+                if (streamId > goAwayFrame.lastStreamId() && isLocalStream(streamId)) {
+                    childChannel.pipeline().fireUserEventTriggered(goAwayFrame.retainedDuplicate());
+                }
+            }
```
I think we need to call goAwayFrame.release()

buchgr replied:
@normanmaurer yep. done!

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2WindowUpdateFrame.java:
```diff
@@ -0,0 +1,22 @@
+
+public interface Http2WindowUpdateFrame extends Http2StreamFrame {
+
+    int windowSizeIncrement();
```
javadocs

netkins commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FrameCodec.java:
```diff
@@ -0,0 +1,266 @@
+            goAway.setLastStreamId(lastStreamId);
+
+            // TODO: Should we also set extraStreamIds?
+            //int extraStreamIds = (lastStreamId - http2Handler.connection().local().lastStreamCreated()) / 2;
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) This block of commented-out lines of code should be removed. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

netkins commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FrameCodec.java:
```diff
@@ -0,0 +1,266 @@
+            final Http2GoAwayFrame goAway = new DefaultHttp2GoAwayFrame(errorCode, debugData);
+            goAway.setLastStreamId(lastStreamId);
+
+            // TODO: Should we also set extraStreamIds?
```
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)

netkins commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FrameCodec.java:
```diff
@@ -0,0 +1,266 @@
+        ctx.fireUserEventTriggered(upgrade.retain());
+        try {
+            Http2Stream stream = http2Handler.connection().stream(Http2CodecUtil.HTTP_UPGRADE_STREAM_ID);
+            // TODO: improve handler/stream lifecycle so that stream isn't active before handler added.
```
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)

netkins commented on codec-http2/src/test/java/io/netty/handler/codec/http2/Http2CodecTest.java:
```diff
@@ -0,0 +1,24 @@
+/**
+ * Unit tests for {@link Http2Codec}.
+ */
+public class Http2CodecTest {
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Add some tests to this class. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS2187)

netkins commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java:
```diff
@@ -28,163 +23,155 @@
+    private void onStreamActive(int streamId) {
+        ChannelFuture future = createStreamChannel(ctx, streamId, streamHandler);
+        Http2StreamChannel childChannel = (Http2StreamChannel) future.channel();
+        Http2StreamChannel prevValue = childChannels.put(streamId, childChannel);
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this useless assignment to local variable "prevValue". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1854)

netkins commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java:
```diff
@@ -28,163 +23,155 @@
+        ChannelFuture future = createStreamChannel(ctx, streamId, streamHandler);
+        Http2StreamChannel childChannel = (Http2StreamChannel) future.channel();
+        Http2StreamChannel prevValue = childChannels.put(streamId, childChannel);
+        // TODO: Probably not assert but proper error handling.
```
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)

netkins commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java:
```diff
@@ -28,163 +23,157 @@
-    public void flush(ChannelHandlerContext ctx) {
-        ctx.flush();
+    private void onStreamActive(int streamId) {
+        ChannelFuture future = createStreamChannel(ctx, streamId, streamHandler);
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this useless assignment to local variable "prevValue". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1854)

netkins commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java:
```diff
@@ -28,163 +23,157 @@
-        ctx.flush();
+    private void onStreamActive(int streamId) {
+        ChannelFuture future = createStreamChannel(ctx, streamId, streamHandler);
+        Http2StreamChannel childChannel = (Http2StreamChannel) future.channel();
```
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)

netkins commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java:
```diff
@@ -28,163 +23,157 @@
+        ChannelFuture future = createStreamChannel(ctx, streamId, streamHandler);
+        Http2StreamChannel childChannel = (Http2StreamChannel) future.channel();
+        Http2StreamChannel prevValue = childChannels.put(streamId, childChannel);
+        // TODO: Probably not assert but proper error handling.
```
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)

netkins commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java:
```diff
@@ -28,163 +23,157 @@
+    private void onStreamActive(int streamId) {
+        ChannelFuture future = createStreamChannel(ctx, streamId, streamHandler);
+        Http2StreamChannel childChannel = (Http2StreamChannel) future.channel();
+        Http2StreamChannel prevValue = childChannels.put(streamId, childChannel);
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this useless assignment to local variable "prevValue". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1854)

netkins commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FrameCodec.java:
```diff
@@ -0,0 +1,285 @@
+        public void onGoAwayReceived(final int lastStreamId, long errorCode, ByteBuf debugData) {
+            final Http2GoAwayFrame goAway = new DefaultHttp2GoAwayFrame(lastStreamId, errorCode, debugData);
+
+            // TODO: Should we also set extraStreamIds?
```
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)

netkins commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FrameCodec.java:
```diff
@@ -0,0 +1,285 @@
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelDuplexHandler;
+import io.netty.channel.ChannelHandler;
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Remove this unused import 'io.netty.channel.ChannelHandler'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

netkins commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FrameCodec.java:
```diff
@@ -0,0 +1,285 @@
+            final Http2GoAwayFrame goAway = new DefaultHttp2GoAwayFrame(lastStreamId, errorCode, debugData);
+
+            // TODO: Should we also set extraStreamIds?
+            //int extraStreamIds = (lastStreamId - http2Handler.connection().local().lastStreamCreated()) / 2;
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) This block of commented-out lines of code should be removed. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

ejona86 commented:
> Should we have the Http2FrameCodec extend from the Http2ConnectionHandler?

I'd favor not. I purposefully didn't do that earlier because I didn't want to leak the connection handler API, which is quite large and not well suited to mixing with the new API.

buchgr commented:
@ejona86 wrote:
> I think I was going to go with something more like a List. That way there wouldn't be any need for a second int → stream mapping in the multiplexing handler. It isn't without problems, but seemed simpler than keeping another map (as a map may require every handler between the multiplexing codec and the frame codec to maintain its own map).

I don't understand. Keep a list of what and where?

> I was thinking later it would be fairly easy to make a new Http2DifferentStream in the frame handler and convert 1:1. We're mainly really only gaining private properties on Http2Stream, but that is used in both directions multiplex codec -> frame codec and frame codec -> multiplex codec. 
> Although, if we're talking about having a map on each stream for properties (likely one per handler) or a map of streams for properties (likely one per handler), I guess it doesn't matter much which we go with. Having a map on each stream is more predictable, so we can use the array-based indexing, and doesn't require noticing stream death. Important properties could also become first-class fields (like maybe the child channel). But both approaches do seem a bit wonky, so I'm open to suggestions and on the fence.

I like that idea. So we could maintain a map of `Http2DifferentStream`s in the frame handler, and have the child channel and so on attached on there. Any other channels could then just access their properties via `frame.stream()`. That might be better than a map in `MultiplexingHandler`. Should we leave it with a map in the multiplexing handler for now, and I address this in a follow up PR?

> I'd favor not. I purposefully didn't do that earlier because I didn't want to leak the connection handler API, which is quite large and not well suited to mixing with the new API.

yeah that's also why I didn't do it in this PR. Didn't want to leak the connection handler. Ok that's settled then.

ejona86 commented:
>> I think I was going to go with something more like a List.

> I don't understand. Keep a list of what and where?

Something like `GoAwayFrame.rejectedStreams() → List`. It's sort of silly, but it would also work.

> Should we leave it with a map in the multiplexing handler for now, and I address this in a follow up PR?

Leaving things with a map for now sounds good. If we want it, it should be easy in a followup PR.

buchgr commented:
> Something like GoAwayFrame.rejectedStreams() → List. It's sort of silly, but it would also work.

Hmm, that might be a good idea actually, and kind of necessary once we remove the map in `Http2MultiplexCodec`. The list returned by `rejectedStreams()` would then be a `Http2DifferentStream` object with the child channel attached as a property.

We just need to be careful with the implementation, as a malicious endpoint could easily send us a GOAWAY frame with say lastStreamId=MAX_INT and we would end up allocating a 16GB array (8 byte * 2 billion) or so, effectively OOMing us.

Scottmitch commented:
> as a malicious endpoint could easily send us a GOAWAY frame with say lastStreamId=MAX_INT

I will take a more detailed look soon ... but just noticed this comment.  Sending a `GO_AWAY` with `lastStreamId=large number` should be expected as this is how graceful shutdown is done:

https://tools.ietf.org/html/rfc7540#section-6.8
> A server that is attempting to gracefully shut down a
   connection SHOULD send an initial GOAWAY frame with the last stream
   identifier set to 2^31-1 and a NO_ERROR code.

ejona86 commented:
> as a malicious endpoint could easily send us a GOAWAY frame with say lastStreamId=MAX_INT

This shouldn't be a problem, because only streams created locally (in the even/odd id sense) will be in the `rejectedStreams()` List. So it isn't really possible to abuse at all remotely. Also it will be bound by `MAX_CONCURRENT_STREAMS` and if that is infinite, will never be more memory than we are already using to keep the stream tree for flow control.

ejona86 commented:
Oh, and I should also mention that the pathological case is when lastStreamId=0, not MAX_INT, because MAX_INT impacts zero streams.

ejona86 commented on codec-http2/src/main/java/io/netty/handler/codec/http2/AbstractHttp2StreamFrame.java:
```diff
@@ -22,17 +22,18 @@
  */
 @UnstableApi
 public abstract class AbstractHttp2StreamFrame implements Http2StreamFrame {
-    private Object stream;
```
Do we want to leave stream as an Object?

buchgr replied:
I had it initially as an `Http2Stream` object but changed it to stream id, cause there is little use for `Http2Stream`. I think eventually we will want a `Http2DifferentStream` object where we can attach properties. I ll make that change in a follow up PR.

ejona86 commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java:
```diff
@@ -67,32 +58,19 @@
 @UnstableApi
-public final class Http2MultiplexCodec extends ChannelDuplexHandler {
-    private static final Http2FrameLogger HTTP2_FRAME_LOGGER = new Http2FrameLogger(INFO, Http2MultiplexCodec.class);
+public class Http2MultiplexCodec extends ChannelDuplexHandler {
```
Why no longer `final`? Maybe you intended to also remove `public`?

ejona86 commented on codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2WindowUpdateFrame.java:
```diff
@@ -0,0 +1,44 @@
+    private final int windowUpdateIncrement;
+
+    public DefaultHttp2WindowUpdateFrame(int windowUpdateIncrement) {
+        this.windowUpdateIncrement = windowUpdateIncrement;
```
Assert positive?

ejona86 commented on codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ResetFrame.java:
```diff
@@ -45,9 +45,14 @@ public DefaultHttp2ResetFrame(long errorCode) {
+
+    @Override
+    public String name() {
+        return "GOAWAY";
```
RST_STREAM?

ejona86 commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Frame.java:
```diff
@@ -19,4 +19,11 @@
+    /**
+     * Returns the name of the HTTP/2 frame e.g. DATA, GOAWAY, etc.
+     */
+    String name();
```
It seems this is used for debugging. I'd expect getClass().getName() to be more reliable and provide more information.

buchgr replied:
I am using it in exception messages. I wanted to mimic the messages of the existing HTTP/2 API, where they use DATA, GOAWAY over class names.

ejona86 commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2HeadersFrame.java:
```diff
@@ -22,8 +22,6 @@
 @UnstableApi
 public interface Http2HeadersFrame extends Http2StreamFrame {
-    @Override
-    Http2HeadersFrame setStream(Object stream);
```
Why remove this?

buchgr replied:
It's now called `setStreamId`.

ejona86 commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java:
```diff
@@ -102,89 +80,130 @@ public Http2MultiplexCodec(boolean server, ChannelHandler streamHandler) {
+        StreamException streamEx = (StreamException) cause;
+        Http2StreamChannel childChannel = childChannels.remove(streamEx.streamId());
+        if (childChannel != null) {
+            childChannel.pipeline().fireExceptionCaught(streamEx);
```
This doesn't close the channel. You need something like in `onStreamClosed0()`

ejona86 commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2StreamFrame.java:
```diff
@@ -19,24 +19,24 @@
     /**
-     * Set the stream identifier for this message.
-     *
-     * @return {@code this}
```
The return documentation is helpful to know what you get back. Otherwise you have to guess.

ejona86 commented:
@buchgr LGTM. One pretty important bug to fix for shutting down the child channel on stream exception. Rest of the stuff is minor. The tests do look easier now. Yay!

netkins commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java:
```diff
@@ -102,89 +80,133 @@ public Http2MultiplexCodec(boolean server, ChannelHandler streamHandler) {
+    private void onStreamActive(int streamId) {
+        ChannelFuture future = createStreamChannel(ctx, streamId, streamHandler);
+        Http2StreamChannel childChannel = (Http2StreamChannel) future.channel();
+        // TODO: Think about error handling in case a stream with the same id already exists in the map.
```
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)

buchgr commented:
@ejona86 PTAL. I fixed the bug and added a test for it https://github.com/netty/netty/pull/5324/files#diff-da759abb652d5db65698f9baf72b9ed2R181

@normanmaurer @Scottmitch I squashed the commits. should be good to go.

buchgr commented:
@normanmaurer @Scottmitch friendly ping :)

ejona86 commented:
FYI I did take another glance (although the squash makes that harder) and it still LGTM

buchgr commented:
@ejona86 sorry about that :(.

normanmaurer commented:
Will check tomorrow!


> On 09 Jun 2016, at 21:10, Jakob Buchgraber <notifications@github.com> wrote:
> 
> @normanmaurer <https://github.com/normanmaurer> @Scottmitch <https://github.com/Scottmitch> friendly ping :)
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/netty/netty/pull/5324#issuecomment-224996206>, or mute the thread <https://github.com/notifications/unsubscribe/AAa0QuXL9mvdsHAMtkqyX5L-rn83aVpmks5qKGUWgaJpZM4Ippk2>.
>

Scottmitch commented:
@buchgr - sorry for the delay ... let me try to clear some time for this today

buchgr commented:
@Scottmitch no problem at all! thanks!

buchgr commented:
friendly ping :)

normanmaurer commented:
@buchgr I am such a slacker 😭 ..... Will look today

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2GoAwayFrame.java:
```diff
@@ -27,6 +27,7 @@
 public final class DefaultHttp2GoAwayFrame extends DefaultByteBufHolder implements Http2GoAwayFrame {
     private final long errorCode;
     private int extraStreamIds;
+    private int lastStreamId = -1;
```
consider cascading all constructors to the newly introduced constructor and just initializing `lastStreamId` in a single place.

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2HeadersFrame.java:
```diff
@@ -63,9 +63,14 @@ public DefaultHttp2HeadersFrame(Http2Headers headers, boolean endStream, int pad
-    public DefaultHttp2HeadersFrame setStream(Object stream) {
-      super.setStream(stream);
-      return this;
+    public DefaultHttp2HeadersFrame setStreamId(int streamId) {
```
what are thoughts about method signatures that expect returning `this` for non-builder objects? They can be difficult to maintain and often require a method be overridden just to return the correct type. I wonder if it is just simpler to just keep these methods `void` ... it may provide the opportunity for more "fluid" code but sometimes this can also be messy. @ejona86 @buchgr @normanmaurer @nmittler - WDYT?

normanmaurer replied:
I actually like to return this... not super hard to get right imho.

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2WindowUpdateFrame.java:
```diff
@@ -0,0 +1,47 @@
+    private final int windowUpdateIncrement;
+
+    public DefaultHttp2WindowUpdateFrame(int windowUpdateIncrement) {
+        if (windowUpdateIncrement < 0) {
```
`ObjectUtil.checkPositive`?

normanmaurer replied:
@Scottmitch actually I think checkPositive will not work as 0 is allowed as well.

Scottmitch replied:
I don't think 0 should be allowed

https://tools.ietf.org/html/rfc7540#section-6.9
>  The payload of a WINDOW_UPDATE frame is one reserved bit plus an
   unsigned 31-bit integer indicating the number of octets that the
   sender can transmit in addition to the existing flow-control window.
   The legal range for the increment to the flow-control window is 1 to
   2^31-1 (2,147,483,647) octets.

> A receiver MUST treat the receipt of a WINDOW_UPDATE frame with an
   flow-control window increment of 0 as a stream error (Section 5.4.2)
   of type PROTOCOL_ERROR; errors on the connection flow-control window
   MUST be treated as a connection error (Section 5.4.1).

normanmaurer replied:
@buchgr @Scottmitch ok then fix it @buchgr  ;)

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Codec.java:
```diff
@@ -0,0 +1,76 @@
+     *
+     * @param server {@code true} this is a server
+     * @param streamHandler the handler added to channels for remotely-created streams. It must be
+     *     {@link ChannelHandler.Sharable}.
```
should we add an assert statement to verify this? the code in `ChannelHandlerAdapter.isSharable()` may be helpful.

ejona86 replied:
I don't mind if we just rely on the Channel's checking. I think that is what ServerBootstrap does.

We could actually relax it and say "generally should be Sharable". If MAX_CONCURRENT_STREAMS is 0-1, technically it is safe :). I don't really care much between any of these various ways.

buchgr replied:
@Scottmitch @ejona86 we are enforcing this in the `Http2MultiplexCodec` constructor.

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FrameCodec.java:
```diff
@@ -0,0 +1,280 @@
+
+    // Override this to signal it will never throw an exception.
+    @Override
+    public void flush(ChannelHandlerContext ctx) {
```
`exceptionCaught` is used in this class and simplifies exception handling ... however does overriding this method provide the same benefits ... doesn't look like it is used in this class ... is it directly called somewhere else?

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2GoAwayFrame.java:
```diff
@@ -46,6 +49,11 @@
     Http2GoAwayFrame setExtraStreamIds(int extraStreamIds);
 
     /**
+     * Returns the last stream identifier if set, or {@code -1} else.
```
clarify the limitation enforced by the [Http2FrameCodec](https://github.com/netty/netty/pull/5324/files#diff-004b9d3b44cadc4c6987434ba0e599beR174) (that this must be `-1` when sending)?

Scottmitch replied:
I now see the comments at the class level (I just jumped to this method in the IDE when I made the comment) ... maybe this is good enough

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FrameCodec.java:
```diff
@@ -0,0 +1,280 @@
+        }
+
+        int lastStreamId = http2Handler.connection().remote().lastStreamCreated()
+                           + frame.extraStreamIds() * 2;
```
can you add a comment as to why `*2`?

Also what is the recommended way to do "graceful shutdown" as described by the rfc https://tools.ietf.org/html/rfc7540#section-6.8:

> A server that is attempting to gracefully shut down a
   connection SHOULD send an initial GOAWAY frame with the last stream
   identifier set to 2^31-1 and a NO_ERROR code

buchgr replied:
@Scottmitch @ejona86 might be a good idea to add a `shutdownGracefully()` or so method to `GoAwayFrame`?

Scottmitch replied:
something like `boolean isGracefulShutdown()`? would this only be meaningful on the sending side ... or would this also return true when ever `lastStreamId() == 2^31-1` on the receiving side? An alternative would be to allow `lastStreamId() == 2^31-1` on the sending side ... and provide a `boolean isGracefulShutdown` argument during construction of the DefaultHttp2GoAwayFrame ... I'm not sure the user will need to know/care if is is a "graceful shutdown" on the receiving side ... the side effects of not being able to creating new streams is the primary effect of this frame.

ejona86 replied:
I'd probably do `extraStreamIds() == MAX_INT`, and we fix the code to saturate. The `extraStreamIds()` was put there to allow graceful shutdown, although I was expecting something more like `extraStreamIds() == 100` just to put an upper bound on the number of new streams.

Maybe `int extraStreamIds()` should be replaced with `boolean relaxedLastStream()`. If `relaxedLastStream()`, then we use MAX_INT, otherwise we use `lastStreamCreated()`. (Other names I considered are strictLastStream and enforceLastStream.)

buchgr replied:
@ejona86 ha! we had the same thought ... I just pushed that :).

Scottmitch replied:
sgtm! @buchgr - Can you expand the comments on the `Http2GoAwayFrame` to include something the the effect that `graceful shutdown as described in the rfc can be accomplished by setting {@code extraStreamIds} to Integer.MAX_VALUE`.

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FrameCodec.java:
```diff
@@ -0,0 +1,280 @@
+        protected void onStreamError(ChannelHandlerContext ctx, Throwable cause,
+                                     Http2Exception.StreamException http2Ex) {
+            try {
+                Http2Stream stream = http2Handler.connection().stream(http2Ex.streamId());
```
instead of using `http2Handler` should we just use `this`? (`http2Handler.connection().stream(http2Ex.streamId())` -> `connection().stream(http2Ex.streamId())`), and make `InternalHttp2ConnectionHandler` a static class? This would simplify the cyclic dependency between this class and the outer class.

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FrameCodec.java:
```diff
@@ -0,0 +1,280 @@
+        @Override
+        public void onRstStreamRead(ChannelHandlerContext ctx, int streamId, long errorCode) {
+            Http2ResetFrame rstFrame = new DefaultHttp2ResetFrame(errorCode);
+            rstFrame.setStreamId(streamId);
```
consider initializing the stream id in the constructor where possible, and only using the `setStreamId()` method where necessary ... maybe a bit more constructor code but it may help clarify the expected stream lifetime and the need for the setter method.

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java:
```diff
@@ -66,32 +57,19 @@
+    private final boolean server;
     private ChannelHandlerContext ctx;
-    private ChannelHandlerContext http2HandlerCtx;
     private volatile Runnable flushTask;
```
not changed in this PR ... but consider not making this volatile ... at worst we will just create a bit more garbage

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java:
```diff
@@ -101,89 +79,132 @@ public Http2MultiplexCodec(boolean server, ChannelHandler streamHandler) {
+            Http2StreamChannel childChannel = childChannels.get(streamId);
+            if (childChannel == null) {
+                ReferenceCountUtil.release(msg);
+                throw new StreamException(streamId, STREAM_CLOSED, format("Received %s frame for an unknown stream %d",
```
We are still fronting this with the `DefaultHttp2ConnectionDecoder ... but if we plan to have this be stand alone eventually maybe add a `TODO` to inherit  [DefaultHttp2ConnectionDecoder.shouldIgnoreHeadersOrDataFrame](https://github.com/netty/netty/blob/4.1/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionDecoder.java#L552) and other "no stream exists" logic.

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java:
```diff
@@ -101,89 +79,132 @@ public Http2MultiplexCodec(boolean server, ChannelHandler streamHandler) {
+            }
+            goAwayFrame.release();
+        } else {
+            throw new UnsupportedMessageTypeException(msg);
```
` ReferenceCountUtil.release(msg);`?

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java:
```diff
@@ -101,89 +79,132 @@ public Http2MultiplexCodec(boolean server, ChannelHandler streamHandler) {
+        }
+    }
+
+    void fireChildReadAndRegister(Http2StreamChannel childChannel, Http2StreamFrame frame) {
```
can this be private?

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java:
```diff
@@ -101,89 +79,132 @@ public Http2MultiplexCodec(boolean server, ChannelHandler streamHandler) {
+    void fireChildReadAndRegister(Http2StreamChannel childChannel, Http2StreamFrame frame) {
+        // Can't use childChannel.fireChannelRead() as it would fire independent of whether
+        // channel.read() had been called.
+        childChannel.fireChildRead(frame);
```
the comment above says we shouldn't do this ... which seems to make sense if the channel hasn't called `read()`?

Scottmitch replied:
actually it looks like `AbstractHttp2StreamChannel.readInProgress` should account for the `read()` being called ... and will queue otherwise ... maybe remove/clarify the above comment?

ejona86 replied:
It is Channel vs Child: fireChannelRead vs fireChildRead. fireChannelRead fires the pipeline directly, IIRC

Scottmitch replied:
ah i see now ... didn't notice th the `child` vs `channel` at first glance.

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java:
```diff
@@ -101,89 +79,132 @@ public Http2MultiplexCodec(boolean server, ChannelHandler streamHandler) {
+        } else if (evt instanceof Http2StreamClosedEvent) {
+            onStreamClosed(streamId);
+        } else {
+            throw new UnsupportedMessageTypeException(evt);
```
`ReferenceCountUtil.release(evt);`?

Scottmitch replied:
maybe just do the release in a try/finally block for all of the above blocks just incase

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java:
```diff
@@ -234,52 +255,10 @@ private void writeFromStreamChannel0(Object msg, boolean flush, ChannelPromise p
-
-    ChannelFuture createStreamChannel(ChannelHandlerContext ctx, Http2Stream stream,
-            ChannelHandler handler) {
+    ChannelFuture createStreamChannel(ChannelHandlerContext ctx, int streamId,
```
can this method be private?

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java:
```diff
@@ -101,89 +79,132 @@ public Http2MultiplexCodec(boolean server, ChannelHandler streamHandler) {
+    private void onStreamClosed0(Http2StreamChannel childChannel) {
+        childChannel.onStreamClosedFired = true;
+        childChannel.fireChildRead(AbstractHttp2StreamChannel.CLOSE_MESSAGE);
+        childChannels.remove(childChannel.streamId);
```
this is already done in `onStreamClosed` ... and also the eventloop in the child channel may be different and cause threading issues if we try to access the map here.

Scottmitch replied:
consider adding `assert childChannel.eventLoop().inEventLoop();` to the start of this method to clarify the thread context

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2WindowUpdateFrame.java:
```diff
@@ -0,0 +1,28 @@
+     * Number of bytes to increment the HTTP/2 stream's or connection's flow control window.
+     */
+    int windowSizeIncrement();
+
```
nit: kill line

Scottmitch commented on codec-http2/src/test/java/io/netty/handler/codec/http2/Http2FrameCodecTest.java:
```diff
@@ -0,0 +1,484 @@
+    @After
+    public void tearDown() {
+        Object o;
+        while ((o = channel.readOutbound()) != null) {
```
just call `channel.finishAndReleaseAll`?

normanmaurer replied:
+1

Scottmitch commented on codec-http2/src/test/java/io/netty/handler/codec/http2/Http2FrameCodecTest.java:
```diff
@@ -0,0 +1,484 @@
+        Http2DataFrame inboundData = inboundHandler.readInbound();
+        assertEquals(releaseLater(new DefaultHttp2DataFrame(bb("hello"), true, 31).setStreamId(stream.id())),
+                     inboundData);
+        assertEquals(1, inboundData.refCnt());
```
does `inboundData` need to be released?

normanmaurer replied:
+1

buchgr replied:
done.

Scottmitch commented on codec-http2/src/test/java/io/netty/handler/codec/http2/Http2FrameCodecTest.java:
```diff
@@ -0,0 +1,484 @@
+        verify(frameWriter).writeData(eq(http2HandlerCtx), eq(1), outboundData.capture(), eq(27),
+                                      eq(true), anyChannelPromise());
+        assertEquals(releaseLater(bb("world")), outboundData.getValue());
+        assertEquals(1, outboundData.getValue().refCnt());
```
does `outboundData.getValue()` need to be released?

buchgr replied:
done

Scottmitch commented on codec-http2/src/test/java/io/netty/handler/codec/http2/Http2FrameCodecTest.java:
```diff
@@ -0,0 +1,484 @@
+        Http2GoAwayFrame expectedFrame = new DefaultHttp2GoAwayFrame(2, Http2Error.NO_ERROR.code(), bb("foo"));
+        Http2GoAwayFrame actualFrame = inboundHandler.readInbound();
+
+        assertEquals(expectedFrame, actualFrame);
```
do we need to release `expectedFrame` (and possibly `actualFrame` too)?

buchgr replied:
done

Scottmitch commented on codec-http2/src/test/java/io/netty/handler/codec/http2/Http2FrameCodecTest.java:
```diff
@@ -0,0 +1,484 @@
+    }
+
+    @Test
+    public void outgoingStreamActiveShouldNotFireUserEvent() throws Exception {
```
this test seems to be verifying that a `Http2StreamActiveEvent` is fired ... consider clarifying the test name.

Scottmitch commented on codec-http2/src/test/java/io/netty/handler/codec/http2/Http2FrameCodecTest.java:
```diff
@@ -0,0 +1,484 @@
+    }
+
+    private static ByteBuf bb(String s) {
+        ByteBuf buf = Unpooled.buffer(s.length() * 4);
```
`return ByteBufUtil.writeUtf8(UnpooledByteBufAllocator.DEFAULT, s);`

Scottmitch commented on codec-http2/src/test/java/io/netty/handler/codec/http2/Http2MultiplexCodecTest.java:
```diff
@@ -12,193 +12,208 @@
+        parentChannel.pipeline().fireChannelRead(dataFrame2);
+
+        assertEquals(headersFrame, inboundHandler.readInbound());
+        assertEquals(dataFrame1, inboundHandler.readInbound());
```
do we need to release these frames?

buchgr replied:
we are only using unpooled buffers in tests. so release should not be necessary?

Scottmitch replied:
IMO we should still release. test code should not be allocator specific if it doesn't need to be.

@normanmaurer - WDYT?

ejona86 replied:
It's required to be released, even for unpooled buffers, otherwise teamcity fails (as it is now) since it has strict leak detection enabled.

buchgr replied:
@ejona86 @Scottmitch ahh okay :). I added release statements to tests.

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/AbstractHttp2StreamFrame.java:
```diff
@@ -22,17 +22,18 @@
     @Override
-    public AbstractHttp2StreamFrame setStream(Object stream) {
-        this.stream = stream;
+    public AbstractHttp2StreamFrame setStreamId(int streamId) {
```
@buchgr do we need todo some validation ? Like streamId can only be positive etc ?

buchgr replied:
@normanmaurer done.

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/AbstractHttp2StreamStateEvent.java:
```diff
@@ -0,0 +1,30 @@
+
+    private final int streamId;
+
+    public AbstractHttp2StreamStateEvent(int streamId) {
```
This is an abstract class... mark the constructor as protected.

normanmaurer replied:
Also again do we need to have some validation for the streamId value ?

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2GoAwayFrame.java:
```diff
@@ -67,6 +68,16 @@ public DefaultHttp2GoAwayFrame(long errorCode, ByteBuf content) {
 
+    DefaultHttp2GoAwayFrame(int lastStreamId, long errorCode, ByteBuf content) {
+        this(errorCode, content);
+        this.lastStreamId = lastStreamId;
```
same question as before... do we need to validate the id ?

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2GoAwayFrame.java:
```diff
@@ -67,6 +68,16 @@ public DefaultHttp2GoAwayFrame(long errorCode, ByteBuf content) {
         this.errorCode = errorCode;
     }
 
+    DefaultHttp2GoAwayFrame(int lastStreamId, long errorCode, ByteBuf content) {
```
Should this be public ?

buchgr replied:
no it's for internal use only.I clarified it in a comment.

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2GoAwayFrame.java:
```diff
@@ -87,6 +98,11 @@ public Http2GoAwayFrame setExtraStreamIds(int extraStreamIds) {
+
+    @Override
     public Http2GoAwayFrame copy() {
         return (Http2GoAwayFrame) super.copy();
```
Actually I think this will not result in the correct thingy... You need to ensure the lastStreamId etc is set as well. Which will not be true if you call super.copy() . Event worse I think this will result in a ClassCastException

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2WindowUpdateFrame.java:
```diff
@@ -0,0 +1,47 @@
+
+    public DefaultHttp2WindowUpdateFrame(int windowUpdateIncrement) {
+        if (windowUpdateIncrement < 0) {
+            throw new IllegalArgumentException("window size increment must be gt zero.");
```
@buchgr should this be ">= 0" ?

Scottmitch commented on codec-http2/src/test/java/io/netty/handler/codec/http2/Http2MultiplexCodecTest.java:
```diff
@@ -12,193 +12,208 @@
+    }
+
+    private static void verifyFramesMultiplexedToCorrectChannel(int streamId, LastInboundHandler inboundHandler,
+                                                         int numFrames) {
```
nit: fix indentation

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2WindowUpdateFrame.java:
```diff
@@ -0,0 +1,47 @@
+/**
+ * The default {@link Http2WindowUpdateFrame} implementation.
+ */
+public class DefaultHttp2WindowUpdateFrame extends AbstractHttp2StreamFrame implements Http2WindowUpdateFrame {
```
@buchgr please remember to put the `@UnstableApi` annotation on all the classes.

buchgr replied:
@normanmaurer if a class inherits from another class/interface that is annotated with `@UnstableApi` ... does that subclass also need to have the annotation?

Scottmitch commented on codec-http2/src/test/java/io/netty/handler/codec/http2/Http2MultiplexCodecTest.java:
```diff
@@ -12,193 +12,208 @@
+            assertNotNull(frame);
+            assertEquals(streamId, frame.streamId());
+        }
+        assertNull(inboundHandler.readInbound());
```
release the frame if it is not null

buchgr replied:
@Scottmitch we are only using unpooled buffers in tests. still release?

Scottmitch commented on codec-http2/src/test/java/io/netty/handler/codec/http2/Http2MultiplexCodecTest.java:
```diff
@@ -12,193 +12,208 @@
+        for (int i = 0; i < numFrames; i++) {
+            Http2StreamFrame frame = inboundHandler.readInbound();
+            assertNotNull(frame);
+            assertEquals(streamId, frame.streamId());
```
release the frame (even if the check fails)?

Scottmitch commented on codec-http2/src/test/java/io/netty/handler/codec/http2/Http2MultiplexCodecTest.java:
```diff
@@ -211,11 +226,30 @@ public void initChannel(Channel channel) {
+            channelActive = false;
+            super.channelInactive(ctx);
+        }
 
```
line 297 ... `writeOutbound` method doesn't appear to be used ... can you remove it?

buchgr replied:
@Scottmitch done

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FrameCodec.java:
```diff
@@ -0,0 +1,280 @@
+        @Override
+        public void onGoAwayReceived(final int lastStreamId, long errorCode, ByteBuf debugData) {
+            final Http2GoAwayFrame goAway = new DefaultHttp2GoAwayFrame(lastStreamId, errorCode, debugData);
+            ctx.fireChannelRead(goAway);
```
just merge both of the above lines

Scottmitch commented:
@buchgr - took a first pass ... its looking good!

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java:
```diff
@@ -101,89 +79,132 @@ public Http2MultiplexCodec(boolean server, ChannelHandler streamHandler) {
+            fireChildReadAndRegister(childChannel, frame);
+        } else if (msg instanceof Http2GoAwayFrame) {
+            Http2GoAwayFrame goAwayFrame = (Http2GoAwayFrame) msg;
+            for (PrimitiveEntry<Http2StreamChannel> entry : childChannels.entries()) {
```
@buchgr can we somehow get rid of creating the Iterator here ?

buchgr replied:
@normanmaurer don't think so, since this is a hash table. getting the keys would also create a new set object. this could should run rarely anyways. once per channel.

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java:
```diff
@@ -101,89 +79,132 @@ public Http2MultiplexCodec(boolean server, ChannelHandler streamHandler) {
+    private void onStreamActive(int streamId) {
+        ChannelFuture future = createStreamChannel(ctx, streamId, streamHandler);
+        Http2StreamChannel childChannel = (Http2StreamChannel) future.channel();
+        childChannels.put(streamId, childChannel);
```
should we assert that this will always return null ?

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2StreamStateEvent.java:
```diff
@@ -0,0 +1,28 @@
+    /**
+     * Returns the HTTP/2 stream identifier for this event.
+     */
+    int streamId();
```
Add some docs that it will return `-1` when not set ?

buchgr replied:
it must always be set.

normanmaurer commented on example/src/main/java/io/netty/example/http2/helloworld/multiplex/server/Http2ServerInitializer.java:
```diff
@@ -45,8 +45,8 @@
-                return new Http2ServerUpgradeCodec(new Http2MultiplexCodec(true,
-                      new HelloWorldHttp2Handler()));
+                return new Http2ServerUpgradeCodec(new Http2Codec(true,
+                                                                  new HelloWorldHttp2Handler()));
```
fix formatting ?

normanmaurer commented on codec-http2/src/test/java/io/netty/handler/codec/http2/Http2MultiplexCodecTest.java:
```diff
@@ -211,11 +226,30 @@ public void initChannel(Channel channel) {
+        return buf;
+    }
+
     static class LastInboundHandler extends ChannelDuplexHandler {
```
final ?

normanmaurer commented on codec-http2/src/test/java/io/netty/handler/codec/http2/Http2MultiplexCodecTest.java:
```diff
@@ -211,11 +226,30 @@ public void initChannel(Channel channel) {
     }
 
+    private static ByteBuf bb(String s) {
+        ByteBuf buf = Unpooled.buffer(s.length() * 4);
```
`ByteBufUtil.utf8MaxBytes(s)` ?

netkins commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FrameCodec.java:
```diff
@@ -0,0 +1,273 @@
+        Http2Connection connection = new DefaultHttp2Connection(server);
+        frameWriter = new Http2OutboundFrameLogger(frameWriter, HTTP2_FRAME_LOGGER);
+        Http2ConnectionEncoder encoder = new DefaultHttp2ConnectionEncoder(connection, frameWriter);
+        Http2FrameReader reader = new Http2InboundFrameLogger(new DefaultHttp2FrameReader(), HTTP2_FRAME_LOGGER);
```
![BLOCKER](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-blocker.png) Close this "DefaultHttp2FrameReader". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS2095)

buchgr replied:
can be ignored as the reader will be closed on channel inactive.

netkins commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java:
```diff
@@ -101,89 +79,141 @@ public Http2MultiplexCodec(boolean server, ChannelHandler streamHandler) {
+    private void onStreamActive(int streamId) {
+        ChannelFuture future = createStreamChannel(ctx, streamId, streamHandler);
+        Http2StreamChannel childChannel = (Http2StreamChannel) future.channel();
+        Http2StreamChannel oldChannel = childChannels.put(streamId, childChannel);
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this useless assignment to local variable "oldChannel". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1854)

buchgr replied:
can be ignored. the local variable is used in the assert one line below.

netkins commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java:
```diff
@@ -101,89 +79,141 @@ public Http2MultiplexCodec(boolean server, ChannelHandler streamHandler) {
+            int streamId = frame.streamId();
+            Http2StreamChannel childChannel = childChannels.get(streamId);
+            if (childChannel == null) {
+                // TODO: Combine with DefaultHttp2ConnectionDecoder.shouldIgnoreHeadersOrDataFrame logic.
```
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)

buchgr replied:
can be ignored.

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/AbstractHttp2StreamStateEvent.java:
```diff
@@ -15,11 +15,17 @@
-    public AbstractHttp2StreamStateEvent(int streamId) {
+    protected AbstractHttp2StreamStateEvent(int streamId) {
+        if (streamId < 0) {
+            throw new IllegalArgumentException("Stream id must be positive.");
```
positive -> non-negative (if zero is allowed).

Scottmitch replied:
also consider using `ObjectUtil.checkPositiveOrZero`

buchgr commented:
@Scottmitch @ejona86 github collapsed our comments regarding graceful shutdown and GOAWAY. I just added overflow checking for last stream id (which seems generally necessary) and so we could do `setExtraStreamIds(MAX_INT)` for graceful shutdown. 

Playing with the code, I am also not convinced that adding a special class or method for this case is necessary.

Scottmitch commented:
@buchgr - I agree ... I pasted a response/request but just do it doesn't get lost here it is...

> sgtm! @buchgr - Can you expand the comments on the `Http2GoAwayFrame` to include something the the effect that `graceful shutdown as described in the rfc can be accomplished by setting {@code extraStreamIds} to Integer.MAX_VALUE`.

buchgr commented:
@Scottmitch I added the comment regarding graceful shutdown, as well as releasing buffers in unit tests.

buchgr commented:
@ejona86 @Scottmitch PTAL ... not sure what the sonarqube failure is all about?

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/AbstractHttp2StreamStateEvent.java:
```diff
@@ -0,0 +1,34 @@
+    private final int streamId;
+
+    protected AbstractHttp2StreamStateEvent(int streamId) {
+        this.streamId = ObjectUtil.checkPositiveOrZero(streamId, "Stream id must be non-negative.");
```
@buchgr actually the second arg is the name of the field and not an exception message ;)

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/AbstractHttp2StreamFrame.java:
```diff
@@ -15,24 +15,29 @@
+        if (this.streamId != -1) {
+            throw new IllegalStateException("Stream identifier may only be set once.");
+        }
+        this.streamId = ObjectUtil.checkPositiveOrZero(streamId, "Stream id must be non-negative.");
```
@buchgr actually the second arg is the name of the field and not an exception message ;)

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2WindowUpdateFrame.java:
```diff
@@ -0,0 +1,49 @@
+    private final int windowUpdateIncrement;
+
+    public DefaultHttp2WindowUpdateFrame(int windowUpdateIncrement) {
+        this.windowUpdateIncrement = checkPositive(windowUpdateIncrement, "window size increment must be gt zero.");
```
@buchgr actually the second arg is the name of the field and not an exception message ;)

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FrameCodec.java:
```diff
@@ -0,0 +1,277 @@
+     */
+    @Override
+    public void handlerRemoved(ChannelHandlerContext ctx) throws Exception {
+        ctx.pipeline().remove(http2Handler);
```
should we also null out the http2Handler ?

buchgr replied:
@normanmaurer do you have a specific reason of why you think we should do that?

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java:
```diff
@@ -66,32 +57,19 @@
-    public Http2MultiplexCodec(boolean server, ChannelHandler streamHandler) {
-        this(server, streamHandler, null);
-    }
+    private final IntObjectMap<Http2StreamChannel> childChannels = new IntObjectHashMap<Http2StreamChannel>();
```
IS there any "good guess" we can make about the capacity and if so supply it ?

buchgr replied:
@normanmaurer no. it's the number of active HTTP/2 streams. it could really be anything. the default initial size is 8.

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java:
```diff
@@ -101,89 +79,141 @@ public Http2MultiplexCodec(boolean server, ChannelHandler streamHandler) {
+            Http2StreamChannel childChannel = childChannels.get(streamEx.streamId());
+            if (childChannel != null) {
+                childChannel.pipeline().fireExceptionCaught(streamEx);
+            }
```
should we log the exception at least if we have no childChannel to propagate it too ?

Scottmitch commented on codec-http2/src/test/java/io/netty/handler/codec/http2/Http2FrameCodecTest.java:
```diff
@@ -0,0 +1,526 @@
+        assertEquals(State.OPEN, stream.state());
+
+        ByteBuf debugData = bb("debug");
+        Http2GoAwayFrame goAwayFrame = new DefaultHttp2GoAwayFrame(Http2Error.NO_ERROR.code(), debugData);
```
do we need to `slice()` this buffer so that the equality comparison below doesn't fail?

buchgr commented:
PTAL

Scottmitch commented:
lgtm plz squash and rebase ... great work!

normanmaurer commented:
+1

> Am 28.06.2016 um 19:53 schrieb Scott Mitchell <notifications@github.com>:
> 
> lgtm plz squash and rebase ... great work!
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

buchgr commented:
@normanmaurer @Scottmitch squashed!

normanmaurer commented:
@buchgr thanks again for this awesome work!

Cherry-picked into 4.1 (3613d15bca15727a4376aeadd9966bc8f4f50a97)
